site/app/controllers/MiscController.php

```diff
@@ -24,59 +24,110 @@ public function run() {
+
+        // if not a zip
+        if (!$is_zip) {
+            // first check that path is valid 
+            foreach (explode(DIRECTORY_SEPARATOR, $path) as $part) {
+                if ($part == ".." || $part == ".") {
+                    throw new \InvalidArgumentException("Cannot have a part of the path just be dots");
+                }
             }
-        }
-        $path = $this->core->getConfig()->getCoursePath();
-        if ($_REQUEST['dir'] === "config_upload") {
-            $check = FileUtils::joinPaths($path, "config_upload");
-            if (!Utils::startsWith($_REQUEST['path'], $check)) {
-                throw new \InvalidArgumentException("Path must start with path to config_upload");
+            $course_path = $this->core->getConfig()->getCoursePath();
+            if ($dir === "config_upload") {
+                $check = FileUtils::joinPaths($course_path, "config_upload");
+                if (!Utils::startsWith($path, $check)) {
+                    throw new \InvalidArgumentException("Path must start with path to config_upload");
+                }
+                if (!file_exists($path)) {
+                    throw new \InvalidArgumentException("File does not exist");
+                }
             }
-            if (!file_exists($_REQUEST['path'])) {
-                throw new \InvalidArgumentException("File does not exist");
+            else if ($dir === "submissions" || $dir === "uploads") {
+                if (!file_exists($path)) {
+                    throw new \InvalidArgumentException("File does not exist");
+                }
             }
-        }
-        else if ($_REQUEST['dir'] === "submissions" || $_REQUEST['dir'] === "uploads") {
-            if (!file_exists($_REQUEST['path'])) {
-                throw new \InvalidArgumentException("File does not exist");
+            else {
+                throw new \InvalidArgumentException("Invalid dir used");
+            }
+            if (!FileUtils::isValidFileName($path)) {
+                throw new \InvalidArgumentException("Not a valid file name");
+            }
+            $folder_names = array();
+            $folder_names[] = "submissions";
+            $folder_names[] = "results";
+            $folder_names[] = "checkout";
+            $folder_names[] = "uploads";
+            $path_anchors = array();
+            $path_anchors[] = FileUtils::joinPaths($course_path, $folder_names[0]);
+            $path_anchors[] = FileUtils::joinPaths($course_path, $folder_names[1]);
+            $path_anchors[] = FileUtils::joinPaths($course_path, $folder_names[2]);
+            $path_anchors[] = FileUtils::joinPaths($course_path, $folder_names[3]);
+            $arr_count = count($path_anchors);
+            $access = false;
+            for ($x = 0; $x < $arr_count; $x++) {
+                if(Utils::startsWith($path, $path_anchors[$x])){
+                    $access = true;
+                }
+            }
```

Comment:
You can use a [foreach](http://php.net/manual/en/control-structures.foreach.php) to iterate through an array. If you needed the index, you could use `$path_anchors as $idx => $path_anchor)`
```
foreach ($path_anchors as $path_anchor) {
    if (Utils::startsWith($path, $path_anchor)) {
        $access = true;
        break;
    }
}
```